File: SellGUI-DynaShop/src/main/java/me/aov/sellgui/dynashop/DynaShopPriceProvider.java
package me.aov.sellgui.dynashop;

import me.aov.sellgui.api.SellGUIPriceProvider;
import me.aov.sellgui.api.SoldItem;
import net.brcdev.shopgui.ShopGuiPlusApi;
import net.brcdev.shopgui.shop.ShopManager.ShopAction;
import net.brcdev.shopgui.shop.item.ShopItem;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.Plugin;

import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;

final class DynaShopPriceProvider implements SellGUIPriceProvider {
    private static final String DYNASHOP_PLUGIN = "ShopGUIPlus-DynaShop";

    private final SellGUIDynaShop plugin;

    DynaShopPriceProvider(SellGUIDynaShop plugin) {
        this.plugin = plugin;
    }

    @Override
    public String getName() {
        return DYNASHOP_PLUGIN;
    }

    @Override
    public int getPriority() {
        return 100;
    }

    @Override
    public boolean isAvailable() {
        return getDynaShopPlugin() != null && Bukkit.getPluginManager().isPluginEnabled("ShopGUIPlus");
    }

    @Override
    public double getSellPrice(Player player, ItemStack itemStack) {
        ShopItem shopItem = findShopItem(player, itemStack);
        if (shopItem == null || !isDynaShopItem(shopItem)) {
            return 0.0;
        }

        int amount = Math.max(1, itemStack.getAmount());
        if (!canSell(shopItem, amount)) {
            return 0.0;
        }

        Object dynamicPrice = loadDynamicPrice(player, shopItem, itemStack);
        if (dynamicPrice == null) {
            return 0.0;
        }

        try {
            double unitPrice;
            if (isProgressivePricing()) {
                double decay = ((Number) dynamicPrice.getClass().getMethod("getDecaySell").invoke(dynamicPrice)).doubleValue();
                unitPrice = ((Number) dynamicPrice.getClass()
                        .getMethod("calculateProgressiveAveragePrice", int.class, double.class, boolean.class)
                        .invoke(dynamicPrice, amount, decay, false)).doubleValue();
            } else {
                unitPrice = ((Number) dynamicPrice.getClass().getMethod("getSellPrice").invoke(dynamicPrice)).doubleValue();
            }
            return unitPrice > 0 && !Double.isNaN(unitPrice) && !Double.isInfinite(unitPrice) ? unitPrice : 0.0;
        } catch (ReflectiveOperationException | RuntimeException ex) {
            debug("Failed to read DynaShop sell price: " + ex.getMessage());
            return 0.0;
        }
    }

    @Override
    public void onItemsSold(Player player, List<SoldItem> soldItems, double totalPrice) {
        for (SoldItem soldItem : soldItems) {
            ItemStack itemStack = soldItem.getItemStack();
            if (itemStack == null) {
                continue;
            }

            ShopItem shopItem = findShopItem(player, itemStack);
            if (shopItem == null || !isDynaShopItem(shopItem)) {
                continue;
            }

            int amount = Math.max(1, soldItem.getAmount());
            if (getSellPrice(player, itemStack) <= 0) {
                continue;
            }

            processSell(player, shopItem, itemStack, amount);
        }
    }

    private ShopItem findShopItem(Player player, ItemStack itemStack) {
        if (itemStack == null) {
            return null;
        }

        try {
            ShopItem shopItem = player != null
                    ? ShopGuiPlusApi.getItemStackShopItem(player, itemStack)
                    : ShopGuiPlusApi.getItemStackShopItem(itemStack);
            if (shopItem != null) {
                return shopItem;
            }
        } catch (Throwable ignored) {
        }

        try {
            return ShopGuiPlusApi.getItemStackShopItem(itemStack);
        } catch (Throwable ignored) {
            return null;
        }
    }

    private boolean isDynaShopItem(ShopItem shopItem) {
        Object shopConfigManager = getShopConfigManager();
        if (shopConfigManager == null) {
            return false;
        }

        String shopId = shopItem.getShop().getId();
        String itemId = shopItem.getId();
        return hasTypeDynaShop(shopConfigManager, shopId, itemId)
                || invokeBoolean(shopConfigManager, "hasDynamicSection", shopId, itemId)
                || invokeBoolean(shopConfigManager, "hasStockSection", shopId, itemId)
                || invokeBoolean(shopConfigManager, "hasRecipeSection", shopId, itemId);
    }

    private boolean hasTypeDynaShop(Object shopConfigManager, String shopId, String itemId) {
        try {
            Method method = shopConfigManager.getClass().getMethod("getItemValue",
                    String.class, String.class, String.class, Class.class);
            Object result = method.invoke(shopConfigManager, shopId, itemId, "typeDynaShop", String.class);
            return result instanceof Optional<?> optional && optional.isPresent();
        } catch (ReflectiveOperationException | RuntimeException ex) {
            debug("Failed to check DynaShop item type: " + ex.getMessage());
            return false;
        }
    }

    private boolean canSell(ShopItem shopItem, int amount) {
        String sellType = getDynaShopType(shopItem, "sell");
        if (!"STOCK".equals(sellType) && !"STATIC_STOCK".equals(sellType)) {
            return true;
        }

        Object dynaShopPlugin = getDynaShopPlugin();
        if (dynaShopPlugin == null) {
            return false;
        }

        try {
            Object priceStock = dynaShopPlugin.getClass().getMethod("getPriceStock").invoke(dynaShopPlugin);
            Object result = priceStock.getClass()
                    .getMethod("canSell", String.class, String.class, int.class)
                    .invoke(priceStock, shopItem.getShop().getId(), shopItem.getId(), amount);
            return Boolean.TRUE.equals(result);
        } catch (ReflectiveOperationException | RuntimeException ex) {
            debug("Failed to check DynaShop stock sell limit: " + ex.getMessage());
            return true;
        }
    }

    private String getDynaShopType(ShopItem shopItem, String action) {
        Object shopConfigManager = getShopConfigManager();
        if (shopConfigManager == null) {
            return "UNKNOWN";
        }

        String shopId = shopItem.getShop().getId();
        String itemId = shopItem.getId();
        try {
            Object type = shopConfigManager.getClass()
                    .getMethod("getRealTypeDynaShop", String.class, String.class, String.class)
                    .invoke(shopConfigManager, shopId, itemId, action);
            if (type != null) {
                return type.toString();
            }
        } catch (ReflectiveOperationException | RuntimeException ignored) {
        }

        try {
            Object type = shopConfigManager.getClass()
                    .getMethod("getTypeDynaShop", String.class, String.class, String.class)
                    .invoke(shopConfigManager, shopId, itemId, action);
            if (type != null && !"NONE".equals(type.toString()) && !"UNKNOWN".equals(type.toString())) {
                return type.toString();
            }
        } catch (ReflectiveOperationException | RuntimeException ignored) {
        }

        try {
            Object type = shopConfigManager.getClass()
                    .getMethod("getTypeDynaShop", String.class, String.class)
                    .invoke(shopConfigManager, shopId, itemId);
            return type == null ? "UNKNOWN" : type.toString();
        } catch (ReflectiveOperationException | RuntimeException ex) {
            return "UNKNOWN";
        }
    }

    private Object loadDynamicPrice(Player player, ShopItem shopItem, ItemStack itemStack) {
        Object listener = getDynaShopListener();
        if (listener == null) {
            return null;
        }

        try {
            Method method = listener.getClass().getMethod("getOrLoadPrice", Player.class, String.class,
                    String.class, ItemStack.class, java.util.Set.class, java.util.Map.class);
            return method.invoke(listener, player, shopItem.getShop().getId(), shopItem.getId(),
                    itemStack, new HashSet<String>(), new HashMap<String, Object>());
        } catch (ReflectiveOperationException | RuntimeException ex) {
            debug("Failed to load DynaShop price: " + ex.getMessage());
            return null;
        }
    }

    private void processSell(Player player, ShopItem shopItem, ItemStack itemStack, int amount) {
        Object listener = getDynaShopListener();
        if (listener == null) {
            return;
        }

        String shopId = shopItem.getShop().getId();
        String itemId = shopItem.getId();
        ItemStack soldStack = itemStack.clone();
        soldStack.setAmount(amount);

        Bukkit.getScheduler().runTaskAsynchronously(plugin, () -> invokeProcessTransaction(listener, shopId, itemId, soldStack, amount));
        invokeUpdateStorage(listener, player, shopId, itemId, amount);
    }

    private void invokeProcessTransaction(Object listener, String shopId, String itemId, ItemStack itemStack, int amount) {
        try {
            Method method = listener.getClass().getDeclaredMethod("processTransactionAsync",
                    String.class, String.class, ItemStack.class, int.class, ShopAction.class);
            method.setAccessible(true);
            method.invoke(listener, shopId, itemId, itemStack, amount, ShopAction.SELL);
        } catch (ReflectiveOperationException | RuntimeException ex) {
            plugin.getLogger().warning("Failed to update DynaShop market transaction for "
                    + shopId + ":" + itemId + ": " + ex.getMessage());
        }
    }

    private void invokeUpdateStorage(Object listener, Player player, String shopId, String itemId, int amount) {
        try {
            Method method = listener.getClass().getDeclaredMethod("updateStorageData",
                    Player.class, String.class, String.class, boolean.class, int.class);
            method.setAccessible(true);
            method.invoke(listener, player, shopId, itemId, false, amount);
        } catch (ReflectiveOperationException | RuntimeException ex) {
            debug("Failed to flush DynaShop storage update: " + ex.getMessage());
        }
    }

    private boolean isProgressivePricing() {
        Object dynaShopPlugin = getDynaShopPlugin();
        if (dynaShopPlugin == null) {
            return false;
        }

        try {
            Object config = dynaShopPlugin.getClass().getMethod("getConfigMain").invoke(dynaShopPlugin);
            Object mode = config.getClass().getMethod("getString", String.class, String.class)
                    .invoke(config, "pricing.calculation-mode", "simple");
            return "progressive".equalsIgnoreCase(String.valueOf(mode));
        } catch (ReflectiveOperationException | RuntimeException ex) {
            return false;
        }
    }

    private boolean invokeBoolean(Object target, String methodName, String shopId, String itemId) {
        try {
            Object result = target.getClass()
                    .getMethod(methodName, String.class, String.class)
                    .invoke(target, shopId, itemId);
            return Boolean.TRUE.equals(result);
        } catch (ReflectiveOperationException | RuntimeException ex) {
            return false;
        }
    }

    private Object getDynaShopListener() {
        Object dynaShopPlugin = getDynaShopPlugin();
        if (dynaShopPlugin == null) {
            return null;
        }

        try {
            return dynaShopPlugin.getClass().getMethod("getDynaShopListener").invoke(dynaShopPlugin);
        } catch (ReflectiveOperationException | RuntimeException ex) {
            debug("Failed to access DynaShop listener: " + ex.getMessage());
            return null;
        }
    }

    private Object getShopConfigManager() {
        Object dynaShopPlugin = getDynaShopPlugin();
        if (dynaShopPlugin == null) {
            return null;
        }

        try {
            return dynaShopPlugin.getClass().getMethod("getShopConfigManager").invoke(dynaShopPlugin);
        } catch (ReflectiveOperationException | RuntimeException ex) {
            debug("Failed to access DynaShop shop config manager: " + ex.getMessage());
            return null;
        }
    }

    private Plugin getDynaShopPlugin() {
        Plugin dynaShop = Bukkit.getPluginManager().getPlugin(DYNASHOP_PLUGIN);
        return dynaShop != null && dynaShop.isEnabled() ? dynaShop : null;
    }

    private void debug(String message) {
        if (plugin.getConfig().getBoolean("debug", false)) {
            plugin.getLogger().info(message);
        }
    }
}

File: SellGUI-DynaShop/src/main/kotlin/me/aov/sellgui/dynashop/DynaShopPriceProvider.kt
package me.aov.sellgui.dynashop

import me.aov.sellgui.api.SellGUIPriceProvider
import me.aov.sellgui.api.SoldItem
import net.brcdev.shopgui.ShopGuiPlusApi
import net.brcdev.shopgui.shop.ShopManager.ShopAction
import net.brcdev.shopgui.shop.item.ShopItem
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.plugin.Plugin
import java.lang.reflect.Method
import java.util.HashMap
import java.util.HashSet
import java.util.Optional

internal class DynaShopPriceProvider(private val plugin: SellGUIDynaShop) : SellGUIPriceProvider {
    override fun getName(): String = DYNASHOP_PLUGIN
    override fun getPriority(): Int = 100
    override fun isAvailable(): Boolean = getDynaShopPlugin() != null && Bukkit.getPluginManager().isPluginEnabled("ShopGUIPlus")

    override fun getSellPrice(player: Player, itemStack: ItemStack): Double {
        val shopItem = findShopItem(player, itemStack) ?: return 0.0
        if (!isDynaShopItem(shopItem)) return 0.0
        val amount = maxOf(1, itemStack.amount)
        if (!canSell(shopItem, amount)) return 0.0
        val dynamicPrice = loadDynamicPrice(player, shopItem, itemStack) ?: return 0.0
        return try {
            val unitPrice = if (isProgressivePricing()) {
                val decay = (dynamicPrice.javaClass.getMethod("getDecaySell").invoke(dynamicPrice) as Number).toDouble()
                (dynamicPrice.javaClass.getMethod("calculateProgressiveAveragePrice", Int::class.javaPrimitiveType, Double::class.javaPrimitiveType, Boolean::class.javaPrimitiveType)
                    .invoke(dynamicPrice, amount, decay, false) as Number).toDouble()
            } else {
                (dynamicPrice.javaClass.getMethod("getSellPrice").invoke(dynamicPrice) as Number).toDouble()
            }
            if (unitPrice > 0 && !unitPrice.isNaN() && !unitPrice.isInfinite()) unitPrice else 0.0
        } catch (exception: ReflectiveOperationException) {
            debug("Failed to read DynaShop sell price: ${exception.message}")
            0.0
        } catch (exception: RuntimeException) {
            debug("Failed to read DynaShop sell price: ${exception.message}")
            0.0
        }
    }

    override fun onItemsSold(player: Player, soldItems: List<SoldItem>, totalPrice: Double) {
        soldItems.forEach { soldItem ->
            val itemStack = soldItem.getItemStack() ?: return@forEach
            val shopItem = findShopItem(player, itemStack) ?: return@forEach
            if (!isDynaShopItem(shopItem)) return@forEach
            val amount = maxOf(1, soldItem.getAmount())
            if (getSellPrice(player, itemStack) <= 0) return@forEach
            processSell(player, shopItem, itemStack, amount)
        }
    }

    private fun findShopItem(player: Player?, itemStack: ItemStack?): ShopItem? {
        itemStack ?: return null
        try {
            val shopItem = if (player != null) ShopGuiPlusApi.getItemStackShopItem(player, itemStack) else ShopGuiPlusApi.getItemStackShopItem(itemStack)
            if (shopItem != null) return shopItem
        } catch (_: Throwable) {
        }
        return try {
            ShopGuiPlusApi.getItemStackShopItem(itemStack)
        } catch (_: Throwable) {
            null
        }
    }

    private fun isDynaShopItem(shopItem: ShopItem): Boolean {
        val manager = getShopConfigManager() ?: return false
        val shopId = shopItem.shop.id
        val itemId = shopItem.id
        return hasTypeDynaShop(manager, shopId, itemId) ||
            invokeBoolean(manager, "hasDynamicSection", shopId, itemId) ||
            invokeBoolean(manager, "hasStockSection", shopId, itemId) ||
            invokeBoolean(manager, "hasRecipeSection", shopId, itemId)
    }

    private fun hasTypeDynaShop(manager: Any, shopId: String, itemId: String): Boolean = try {
        val result = manager.javaClass.getMethod("getItemValue", String::class.java, String::class.java, String::class.java, Class::class.java)
            .invoke(manager, shopId, itemId, "typeDynaShop", String::class.java)
        result is Optional<*> && result.isPresent
    } catch (exception: ReflectiveOperationException) {
        debug("Failed to check DynaShop item type: ${exception.message}")
        false
    } catch (exception: RuntimeException) {
        debug("Failed to check DynaShop item type: ${exception.message}")
        false
    }

    private fun canSell(shopItem: ShopItem, amount: Int): Boolean {
        val sellType = getDynaShopType(shopItem, "sell")
        if (sellType != "STOCK" && sellType != "STATIC_STOCK") return true
        val dynaShop = getDynaShopPlugin() ?: return false
        return try {
            val stock = dynaShop.javaClass.getMethod("getPriceStock").invoke(dynaShop)
            stock.javaClass.getMethod("canSell", String::class.java, String::class.java, Int::class.javaPrimitiveType)
                .invoke(stock, shopItem.shop.id, shopItem.id, amount) == true
        } catch (exception: ReflectiveOperationException) {
            debug("Failed to check DynaShop stock sell limit: ${exception.message}")
            true
        } catch (exception: RuntimeException) {
            debug("Failed to check DynaShop stock sell limit: ${exception.message}")
            true
        }
    }

    private fun getDynaShopType(shopItem: ShopItem, action: String): String {
        val manager = getShopConfigManager() ?: return "UNKNOWN"
        val shopId = shopItem.shop.id
        val itemId = shopItem.id
        try {
            manager.javaClass.getMethod("getRealTypeDynaShop", String::class.java, String::class.java, String::class.java)
                .invoke(manager, shopId, itemId, action)?.let { return it.toString() }
        } catch (_: ReflectiveOperationException) {
        } catch (_: RuntimeException) {
        }
        try {
            manager.javaClass.getMethod("getTypeDynaShop", String::class.java, String::class.java, String::class.java)
                .invoke(manager, shopId, itemId, action)?.toString()?.takeIf { it != "NONE" && it != "UNKNOWN" }?.let { return it }
        } catch (_: ReflectiveOperationException) {
        } catch (_: RuntimeException) {
        }
        return try {
            manager.javaClass.getMethod("getTypeDynaShop", String::class.java, String::class.java).invoke(manager, shopId, itemId)?.toString() ?: "UNKNOWN"
        } catch (_: ReflectiveOperationException) {
            "UNKNOWN"
        } catch (_: RuntimeException) {
            "UNKNOWN"
        }
    }

    private fun loadDynamicPrice(player: Player, shopItem: ShopItem, itemStack: ItemStack): Any? {
        val listener = getDynaShopListener() ?: return null
        return try {
            listener.javaClass.getMethod("getOrLoadPrice", Player::class.java, String::class.java, String::class.java, ItemStack::class.java, java.util.Set::class.java, java.util.Map::class.java)
                .invoke(listener, player, shopItem.shop.id, shopItem.id, itemStack, HashSet<String>(), HashMap<String, Any>())
        } catch (exception: ReflectiveOperationException) {
            debug("Failed to load DynaShop price: ${exception.message}")
            null
        } catch (exception: RuntimeException) {
            debug("Failed to load DynaShop price: ${exception.message}")
            null
        }
    }

    private fun processSell(player: Player, shopItem: ShopItem, itemStack: ItemStack, amount: Int) {
        val listener = getDynaShopListener() ?: return
        val shopId = shopItem.shop.id
        val itemId = shopItem.id
        val soldStack = itemStack.clone().also { it.amount = amount }
        Bukkit.getScheduler().runTaskAsynchronously(plugin, Runnable { invokeProcessTransaction(listener, shopId, itemId, soldStack, amount) })
        invokeUpdateStorage(listener, player, shopId, itemId, amount)
    }

    private fun invokeProcessTransaction(listener: Any, shopId: String, itemId: String, itemStack: ItemStack, amount: Int) {
        try {
            listener.javaClass.getDeclaredMethod("processTransactionAsync", String::class.java, String::class.java, ItemStack::class.java, Int::class.javaPrimitiveType, ShopAction::class.java)
                .also { it.isAccessible = true }
                .invoke(listener, shopId, itemId, itemStack, amount, ShopAction.SELL)
        } catch (exception: ReflectiveOperationException) {
            plugin.logger.warning("Failed to update DynaShop market transaction for $shopId:$itemId: ${exception.message}")
        } catch (exception: RuntimeException) {
            plugin.logger.warning("Failed to update DynaShop market transaction for $shopId:$itemId: ${exception.message}")
        }
    }

    private fun invokeUpdateStorage(listener: Any, player: Player, shopId: String, itemId: String, amount: Int) {
        try {
            listener.javaClass.getDeclaredMethod("updateStorageData", Player::class.java, String::class.java, String::class.java, Boolean::class.javaPrimitiveType, Int::class.javaPrimitiveType)
                .also { it.isAccessible = true }
                .invoke(listener, player, shopId, itemId, false, amount)
        } catch (exception: ReflectiveOperationException) {
            debug("Failed to flush DynaShop storage update: ${exception.message}")
        } catch (exception: RuntimeException) {
            debug("Failed to flush DynaShop storage update: ${exception.message}")
        }
    }

    private fun isProgressivePricing(): Boolean {
        val dynaShop = getDynaShopPlugin() ?: return false
        return try {
            val config = dynaShop.javaClass.getMethod("getConfigMain").invoke(dynaShop)
            "progressive".equals(config.javaClass.getMethod("getString", String::class.java, String::class.java).invoke(config, "pricing.calculation-mode", "simple")?.toString(), ignoreCase = true)
        } catch (_: ReflectiveOperationException) {
            false
        } catch (_: RuntimeException) {
            false
        }
    }

    private fun invokeBoolean(target: Any, methodName: String, shopId: String, itemId: String): Boolean = try {
        target.javaClass.getMethod(methodName, String::class.java, String::class.java).invoke(target, shopId, itemId) == true
    } catch (_: ReflectiveOperationException) {
        false
    } catch (_: RuntimeException) {
        false
    }

    private fun getDynaShopListener(): Any? {
        val dynaShop = getDynaShopPlugin() ?: return null
        return try { dynaShop.javaClass.getMethod("getDynaShopListener").invoke(dynaShop) } catch (exception: ReflectiveOperationException) { debug("Failed to access DynaShop listener: ${exception.message}"); null } catch (exception: RuntimeException) { debug("Failed to access DynaShop listener: ${exception.message}"); null }
    }

    private fun getShopConfigManager(): Any? {
        val dynaShop = getDynaShopPlugin() ?: return null
        return try { dynaShop.javaClass.getMethod("getShopConfigManager").invoke(dynaShop) } catch (exception: ReflectiveOperationException) { debug("Failed to access DynaShop shop config manager: ${exception.message}"); null } catch (exception: RuntimeException) { debug("Failed to access DynaShop shop config manager: ${exception.message}"); null }
    }

    private fun getDynaShopPlugin(): Plugin? = Bukkit.getPluginManager().getPlugin(DYNASHOP_PLUGIN)?.takeIf { it.isEnabled }
    private fun debug(message: String) { if (plugin.config.getBoolean("debug", false)) plugin.logger.info(message) }

    private companion object { const val DYNASHOP_PLUGIN = "ShopGUIPlus-DynaShop" }
}
